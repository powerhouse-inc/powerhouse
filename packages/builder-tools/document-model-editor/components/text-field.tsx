import { zodResolver } from "@hookform/resolvers/zod";
import {
  forwardRef,
  useCallback,
  useEffect,
  useId,
  useImperativeHandle,
  useRef,
} from "react";
import type { FieldValues } from "react-hook-form";
import { useForm } from "react-hook-form";
import { z } from "zod";
import { createNameSchema } from "../schemas/inputs.js";
import { compareStringsWithoutWhitespace } from "../utils/helpers.js";
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "./form.js";
import type { TextareaHandle } from "./text-area.js";
import { Textarea } from "./text-area.js";

type TextFieldProps = {
  name: string;
  value: string | null | undefined;
  onSubmit: (value: string) => void;
  label?: string;
  placeholder?: string;
  className?: string;
  rows?: number;
  focusOnMount?: boolean;
  required?: boolean;
  allowEmpty?: boolean;
  unique?: string[];
  shouldReset?: boolean;
  onChange?: (value: string) => void;
};

type TextFieldHandle = {
  focus: () => void;
};

export const TextField = forwardRef<TextFieldHandle, TextFieldProps>(
  (
    {
      name,
      value,
      onSubmit,
      label,
      placeholder,
      unique,
      className = "",
      rows = 1,
      focusOnMount = false,
      required = false,
      allowEmpty = false,
      shouldReset = false,
      onChange,
    },
    ref,
  ) => {
    const textareaRef = useRef<TextareaHandle | null>(null);
    const id = useId();
    useEffect(() => {
      if (focusOnMount && textareaRef.current) {
        textareaRef.current.focus();
      }
    }, [focusOnMount]);

    const fieldSchema = z.object({
      [name]: createNameSchema({ required, allowEmpty, unique }),
    });

    const form = useForm({
      resolver: zodResolver(fieldSchema),
      defaultValues: {
        [name]: value ?? "",
      },
    });

    const handleSubmit = useCallback(
      (values: FieldValues) => {
        const newValue = values[name] as string | undefined;
        if (newValue === undefined || value === newValue) return;
        onSubmit(newValue);
        if (shouldReset) form.reset({ [name]: "" });
      },
      [name, value, onSubmit, form, shouldReset],
    );

    const handleBlur = useCallback(async () => {
      const currentValue = form.getValues()[name] ?? "";

      if (value === null || value === undefined) {
        if (!currentValue || currentValue.trim() === "") return;
      }

      if (compareStringsWithoutWhitespace(currentValue, value ?? "")) return;

      try {
        await form.trigger(); // Trigger validation
        if (form.formState.isValid) {
          await form.handleSubmit(handleSubmit)();
        }
      } catch (e) {
        // Allow blur to proceed even if validation fails
      }
    }, [form, handleSubmit, name, value]);

    const onEnterKeyDown = useCallback(
      (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
        if (e.key === "Enter") {
          e.preventDefault();
          (e.target as HTMLTextAreaElement).blur();
        }
      },
      [],
    );

    const handleChange = useCallback(
      (e: React.ChangeEvent<HTMLTextAreaElement>) => {
        const newValue = e.target.value;
        onChange?.(newValue);
      },
      [onChange],
    );

    useImperativeHandle(ref, () => ({
      focus: () => textareaRef.current?.focus(),
    }));

    useEffect(() => {
      form.reset({ [name]: value ?? "" });
    }, [form, name, value]);

    return (
      <Form {...form}>
        <FormField
          control={form.control}
          name={name}
          render={({ field }) => (
            <FormItem className="grid h-full grid-rows-[auto,1fr] gap-2 overflow-visible">
              {!!label && (
                <FormLabel
                  htmlFor={name}
                  className="text-sm font-medium text-gray-700"
                >
                  {label}
                </FormLabel>
              )}
              <FormControl>
                <Textarea
                  {...field}
                  id={id}
                  name={name}
                  ref={(node) => {
                    if (node) {
                      field.ref(node.element);
                      textareaRef.current = node;
                    }
                  }}
                  placeholder={placeholder}
                  onBlur={handleBlur}
                  onChange={(e) => {
                    field.onChange(e);
                    handleChange(e);
                  }}
                  onKeyDown={onEnterKeyDown}
                  rows={rows}
                  className={className}
                />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
      </Form>
    );
  },
);

TextField.displayName = "TextField";
