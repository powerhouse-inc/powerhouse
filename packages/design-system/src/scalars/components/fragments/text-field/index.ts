export { TextField, type TextFieldProps } from "./text-field.js";
