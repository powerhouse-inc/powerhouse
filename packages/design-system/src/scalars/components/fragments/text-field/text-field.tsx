import {
  TextInput,
  type TextInputProps,
} from "../../../../ui/components/data-entry/text-input/text-input.js";
import type { FieldErrorHandling, WithDifference } from "../../types.js";
import { withFieldValidation } from "../with-field-validation/index.js";

export type TextFieldProps = TextInputProps &
  FieldErrorHandling &
  WithDifference<string>;

export const TextField = withFieldValidation<TextFieldProps>(TextInput);

TextField.displayName = "TextField";
