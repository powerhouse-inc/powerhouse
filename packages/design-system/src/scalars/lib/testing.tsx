import { render, type RenderResult } from "@testing-library/react";
import { Form } from "../components/index.js";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
