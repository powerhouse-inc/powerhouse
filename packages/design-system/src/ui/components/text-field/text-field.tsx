import type {
  FieldErrorHandling,
  WithDifference,
} from "@powerhousedao/design-system/ui";
import type { TextInputProps } from "@powerhousedao/design-system/ui";
import {
  TextInput,
  withFieldValidation,
} from "@powerhousedao/design-system/ui";

export type TextFieldProps = TextInputProps &
  FieldErrorHandling &
  WithDifference<string>;

export const TextField = withFieldValidation<TextFieldProps>(TextInput);

TextField.displayName = "TextField";
