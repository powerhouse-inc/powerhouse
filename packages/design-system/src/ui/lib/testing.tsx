import { Form } from "@powerhousedao/design-system";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
