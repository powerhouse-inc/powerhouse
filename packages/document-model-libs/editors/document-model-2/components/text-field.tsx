import {
  useCallback,
  useEffect,
  useRef,
  forwardRef,
  useImperativeHandle,
} from "react";
import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { z } from "zod";
import {
  Form,
  FormField,
  FormItem,
  FormControl,
  FormMessage,
  FormLabel,
} from "./form";
import { Textarea, TextareaHandle } from "./text-area";
import { createNameSchema } from "../schemas";
import { compareStringsWithoutWhitespace } from "../utils/helpers";

type TextFieldProps = {
  name: string;
  value: string | null | undefined;
  onSubmit: (value: string) => void;
  label?: string;
  placeholder?: string;
  className?: string;
  rows?: number;
  focusOnMount?: boolean;
  required?: boolean;
  allowEmpty?: boolean;
  unique?: string[];
  shouldReset?: boolean;
  onChange?: (value: string) => void;
};

type TextFieldHandle = {
  focus: () => void;
};

export const TextField = forwardRef<TextFieldHandle, TextFieldProps>(
  (
    {
      name,
      value,
      onSubmit,
      label,
      placeholder,
      unique,
      className = "",
      rows = 1,
      focusOnMount = false,
      required = false,
      allowEmpty = false,
      shouldReset = false,
      onChange,
    },
    ref,
  ) => {
    const textareaRef = useRef<TextareaHandle | null>(null);

    useEffect(() => {
      if (focusOnMount && textareaRef.current) {
        textareaRef.current.focus();
      }
    }, [focusOnMount]);

    const fieldSchema = z.object({
      [name]: createNameSchema({ required, allowEmpty, unique }),
    });

    type FieldValues = z.infer<typeof fieldSchema>;

    const form = useForm<FieldValues>({
      resolver: zodResolver(fieldSchema),
      defaultValues: {
        [name]: value ?? "",
      },
    });

    const handleSubmit = useCallback(
      (values: FieldValues) => {
        const newValue = values[name as keyof FieldValues];
        if (newValue === undefined || value === newValue) return;
        onSubmit(newValue);
        if (shouldReset) form.reset({ [name]: "" });
      },
      [name, value, onSubmit, form, shouldReset],
    );

    const handleBlur = useCallback(async () => {
      const currentValue = form.getValues()[name] ?? "";

      if (value === null || value === undefined) {
        if (!currentValue || currentValue.trim() === "") return;
      }

      if (compareStringsWithoutWhitespace(currentValue, value ?? "")) return;

      try {
        await form.trigger(); // Trigger validation
        if (form.formState.isValid) {
          await form.handleSubmit(handleSubmit)();
        }
      } catch (e) {
        // Allow blur to proceed even if validation fails
      }
    }, [form, handleSubmit, name, value]);

    const onEnterKeyDown = useCallback(
      (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
        if (e.key === "Enter") {
          e.preventDefault();
          (e.target as HTMLTextAreaElement).blur();
        }
      },
      [],
    );

    const handleChange = useCallback(
      (e: React.ChangeEvent<HTMLTextAreaElement>) => {
        const newValue = e.target.value;
        onChange?.(newValue);
      },
      [onChange],
    );

    useImperativeHandle(ref, () => ({
      focus: () => textareaRef.current?.focus(),
    }));

    useEffect(() => {
      form.reset({ [name]: value ?? "" });
    }, [form, name, value]);

    return (
      <Form {...form}>
        <FormField
          control={form.control}
          name={name}
          render={({ field }) => (
            <FormItem className="grid h-full grid-rows-[auto,1fr] gap-2 overflow-visible">
              {!!label && (
                <FormLabel
                  htmlFor={name}
                  className="text-sm font-medium text-gray-700"
                >
                  {label}
                </FormLabel>
              )}
              <FormControl>
                <Textarea
                  {...field}
                  id={name}
                  ref={(node) => {
                    if (node) {
                      field.ref(node.element);
                      textareaRef.current = node;
                    }
                  }}
                  placeholder={placeholder}
                  onBlur={handleBlur}
                  onChange={(e) => {
                    field.onChange(e);
                    handleChange(e);
                  }}
                  onKeyDown={onEnterKeyDown}
                  rows={rows}
                  className={className}
                />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
      </Form>
    );
  },
);

TextField.displayName = "TextField";
