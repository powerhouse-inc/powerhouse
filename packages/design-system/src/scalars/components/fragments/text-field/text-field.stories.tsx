import type { Meta, StoryObj } from "@storybook/react";
import { TextField } from "./text-field";

const meta = {
  title: "Document Engineering/Fragments/TextField",
  component: TextField,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {
    label: {
      control: "text",
    },
    description: {
      control: "text",
    },
    value: {
      control: "text",
    },
    required: {
      control: "boolean",
    },
    disabled: {
      control: "boolean",
    },
    placeholder: {
      control: "text",
    },
    errors: {
      control: "object",
    },
    warnings: {
      control: "object",
    },
  },
} satisfies Meta<typeof TextField>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    label: "Text Input",
    placeholder: "Type something...",
  },
};

export const WithDescription: Story = {
  args: {
    label: "Username",
    description: "You will need this to log in to your account.",
    placeholder: "johndoe",
  },
};

export const Required: Story = {
  args: {
    label: "Username",
    required: true,
    placeholder: "johndoe",
  },
};

export const WithValue: Story = {
  args: {
    label: "First Name",
    value: "John",
  },
};

export const Disabled: Story = {
  args: {
    label: "Read Only Field",
    value: "This field is disabled",
    disabled: true,
  },
};

export const WithError: Story = {
  args: {
    label: "Username",
    value: "jo",
    required: true,
    errors: ["Username must be at least 5 characters long"],
  },
};

export const WithWarning: Story = {
  args: {
    label: "Username",
    value: "LKlfnwuirenfanmdpmawef",
    warnings: ["Will you remember this?"],
  },
};

export const WithDefaultValue: Story = {
  args: {
    label: "Username",
    default: "johndoe",
    placeholder: "Enter username",
  },
};
