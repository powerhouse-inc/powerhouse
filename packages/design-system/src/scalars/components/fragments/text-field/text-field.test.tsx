import { screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { renderWithForm } from "../../../lib/testing.js";
import { TextField } from "./text-field.js";

describe("TextField", () => {
  it("should match snapshot", () => {
    const { container } = renderWithForm(
      <TextField name="test" label="Test Label" />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should render with label", () => {
    renderWithForm(<TextField name="field" label="Field Label" />);
    expect(screen.getByText("Field Label")).toBeInTheDocument();
  });

  it("should show required indicator when required", () => {
    renderWithForm(<TextField name="name" label="Name" required />);
    expect(screen.getByText("*")).toBeInTheDocument();
  });

  it("should render with initial value", () => {
    renderWithForm(
      <TextField
        name="name"
        label="Name"
        value="John Doe"
        onChange={() => null}
      />,
    );
    expect(screen.getByDisplayValue("John Doe")).toBeInTheDocument();
  });

  it("should render with default value", () => {
    renderWithForm(
      <TextField name="name" label="Name" defaultValue="John Doe" />,
    );
    expect(screen.getByDisplayValue("John Doe")).toBeInTheDocument();
  });

  it("should render in disabled state", () => {
    renderWithForm(<TextField name="name" label="Name" disabled />);
    expect(screen.getByRole("textbox")).toBeDisabled();
  });

  it("should display error messages", async () => {
    renderWithForm(
      <TextField name="name" label="Name" errors={["Name is required"]} />,
    );
    await waitFor(() =>
      expect(screen.getByText("Name is required")).toBeInTheDocument(),
    );
  });

  it("should display warning messages", () => {
    renderWithForm(
      <TextField name="name" label="Name" warnings={["Name is too long"]} />,
    );
    expect(screen.getByText("Name is too long")).toBeInTheDocument();
  });

  it("should allow user input", async () => {
    const user = userEvent.setup();
    renderWithForm(<TextField name="name" label="Name" />);

    const input = screen.getByRole("textbox");
    await user.type(input, "John Doe");

    expect(input).toHaveValue("John Doe");
  });

  it("should display character counter when maxLength is provided", () => {
    renderWithForm(
      <TextField
        name="name"
        label="Name"
        value="John"
        maxLength={10}
        onChange={() => null}
      />,
    );
    expect(screen.getByText("4")).toBeInTheDocument();
    expect(screen.getByText("/10")).toBeInTheDocument();
  });

  it("should display character counter with empty value when maxLength is provided", () => {
    renderWithForm(
      <TextField
        name="name"
        label="Name"
        value=""
        maxLength={11}
        onChange={() => null}
      />,
    );
    expect(screen.getByText("0")).toBeInTheDocument();
    expect(screen.getByText("/11")).toBeInTheDocument();
  });

  it("should enable browser autocomplete when autoComplete is true", () => {
    renderWithForm(<TextField name="name" label="Name" autoComplete={true} />);
    expect(screen.getByRole("textbox")).toHaveAttribute("autocomplete", "on");
  });

  it("should disable browser autocomplete when autoComplete is false", () => {
    renderWithForm(<TextField name="name" label="Name" autoComplete={false} />);
    expect(screen.getByRole("textbox")).toHaveAttribute("autocomplete", "off");
  });

  it("should not set autocomplete attribute when autoComplete is undefined", () => {
    renderWithForm(<TextField name="name" label="Name" />);
    expect(screen.getByRole("textbox")).not.toHaveAttribute("autocomplete");
  });

  it("should enable browser spellcheck when spellCheck is true", () => {
    renderWithForm(<TextField name="name" label="Name" spellCheck={true} />);
    expect(screen.getByRole("textbox")).toHaveAttribute("spellcheck", "true");
  });

  it("should disable browser spellcheck when spellCheck is false", () => {
    renderWithForm(<TextField name="name" label="Name" spellCheck={false} />);
    expect(screen.getByRole("textbox")).toHaveAttribute("spellcheck", "false");
  });

  it("should not set spellcheck attribute when spellCheck is undefined", () => {
    renderWithForm(<TextField name="name" label="Name" />);
    expect(screen.getByRole("textbox")).not.toHaveAttribute("spellcheck");
  });
});
