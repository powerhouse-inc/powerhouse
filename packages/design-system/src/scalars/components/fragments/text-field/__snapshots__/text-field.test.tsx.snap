// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextField > should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="inline-flex items-center text-sm font-medium leading-[22px] text-gray-900 group-hover:text-gray-900 dark:text-gray-400 dark:group-hover:text-slate-50"
      role="label"
    >
      Test Label
    </label>
    <input
      class="flex h-10 w-full rounded-md text-sm text-gray-900 border border-gray-300 bg-background px-3 py-2 placeholder:text-gray-600 ring-offset-background focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-1 disabled:cursor-not-allowed disabled:border-gray-200 disabled:text-gray-500"
      value=""
    />
  </div>
</div>
`;
